ekncontent/ekncontent/eknc-engine.c

```diff
@@ -0,0 +1,499 @@
+                          const GValue *value,
+                          GParamSpec *pspec)
+{
+  EkncEngine *self = EKNC_ENGINE (object);
+
+  switch (prop_id)
+    {
+    case PROP_DEFAULT_APP_ID:
+      g_clear_pointer (&self->default_app_id, g_free);
+      self->default_app_id = g_value_dup_string (value);
+      break;
+
+    case PROP_LANGUAGE:
+      g_clear_pointer (&self->language, g_free);
+      self->language = g_value_dup_string (value);
+      g_object_set (G_OBJECT (self->xapian_bridge), "language", self->language, NULL);
+      break;
+
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+    }
+}
+
+static void
+eknc_engine_finalize (GObject *object)
+{
+  EkncEngine *self = EKNC_ENGINE (object);
+
+  g_clear_pointer (&self->default_app_id, g_free);
+  g_clear_pointer (&self->language, g_free);
+  g_clear_object (&self->xapian_bridge);
+  g_clear_object (&self->domains);
+}
+
+static void
+eknc_engine_class_init (EkncEngineClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->get_property = eknc_engine_get_property;
+  object_class->set_property = eknc_engine_set_property;
+  object_class->finalize = eknc_engine_finalize;
+
+  /**
+   * EkncEngine:default-app-id:
+   *
+   * The default application the engine should serve up content for.
+   */
+  eknc_engine_props[PROP_DEFAULT_APP_ID] =
+    g_param_spec_string ("default-app-id", "Default Application ID",
+      "Default Application ID",
+      "", G_PARAM_READWRITE | G_PARAM_CONSTRUCT | G_PARAM_STATIC_STRINGS);
+
+  /**
+   * EkncEngine:language:
+   *
+   * The ISO639 language code which will be used for various search features,
+   * such as term stemming and spelling correction.
+   */
+  eknc_engine_props[PROP_LANGUAGE] =
+    g_param_spec_string ("language", "Language",
+      "The language to use",
+      "", G_PARAM_READWRITE | G_PARAM_CONSTRUCT | G_PARAM_STATIC_STRINGS);
+
+  g_object_class_install_properties (object_class,
+                                     NPROPS,
+                                     eknc_engine_props);
+}
+
+static void
+eknc_engine_init (EkncEngine *self)
+{
+  self->xapian_bridge = g_object_new (EKNC_TYPE_XAPIAN_BRIDGE,
+                                      "language", self->language,
+                                       NULL);
+  self->domains = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_object_unref);
+}
+
+/**
+ * eknc_engine_test_link:
+ * @self: the engine
+ * @link: the ekn id of link to check for
+ * @error: #GError for error reporting.
+ *
+ * Attempts to determine if the given link corresponds to content within
+ * the default domain.
+ *
+ * Returns: Returns an EKN URI to that content if so, and NULL otherwise.
+ */
+const gchar *
+eknc_engine_test_link (EkncEngine *self,
+                       const gchar *link,
+                       GError **error)
+{
+  g_return_val_if_fail (EKNC_IS_ENGINE (self), NULL);
```

Comment:
Added a lot of `g_return_val_if_fail (err == NULL || *err == NULL, NULL);` check though